client/state/data-layer/wpcom-http/README.md

```diff
@@ -0,0 +1,140 @@
+This `UNLIKE_POST` action will end up rolling back the first like, which in effect could not be completed because the server failed to update.
+
+
+### Step D: Reducers see action and update app state
+
+The actual `LIKE_POST` action (and possible the `UNLIKE_POST` as well) hit the reducers and the app state is updated.
+This will trigger a re-render on the post component which started the chain and we come full circle.
+
+## Implementation
+
+If you would like to take advantage of this system you will want to use the provided helper methods and structure your code into three pieces: a function which generates the descriptive HTTP requests; a function which handles successful responses; and a function which handles failing responses.
+
+Each of these functions will take the normal middleware arguments but the success and failure functions take an additional argument which is the response data or error respectively.
+
+```js
+// API Middleware, Post Like
+import { LIKE_POST } from 'state/action-types';
+import { http } from 'state/data-layer/wpcom-http/actions';
+import { QUEUE_REQUEST } from 'state/data-layer/wpcom-http/constants';
+
+/**
+ * @see https://developer.wordpress.com/docs/api/1.1/post/sites/%24site/posts/%24post_ID/likes/new/ API description
+ */
+const likePost = ( { dispatch }, action, next ) => {
+	// dispatch intent to issue HTTP request
+	dispatch( http( {
+		method: 'POST',
+		path: '/sites/%s/posts/%d/likes/new',
+		pathArgs: [ action.siteId, action.postId ],
+		
+		// we can reuse the original action because
+		// we will use meta to indicate which part
+		// of the process we are in
+		onSuccess: action,
+		onFailure: action,
+		
+		// indicate what should happen if we have
+		// no network connection: queue to replay
+		// when the network reconnects
+		// (not implemented yet)
+		whenOffline: QUEUE_REQUEST,
+	} ) );
+	
+	// feed LIKE_POST action along to reducers
+	// and skip additional data-layer middleware
+	next( action );
+}
+
+/**
+ * Called on success from HTTP middleware with `data` parameter
+ *
+ * This is the place to map fromAPI to Calypso formats
+ */
+const verifyLike = ( { dispatch }, { siteId, postId }, next, data ) => {
+	// this is a response to data coming in from the data layer,
+	// so skip further data-layer middleware with next vs. dispatch
+	next( {
+		type: data.i_like ? LIKE_POST : UNLIKE_POST,
+		siteId,
+		postId,
+		likeCount: data.like_count,
+	} );
+}
+
+/**
+ * Called on failure from the HTTP middleware with `error` parameter
+ */
+const undoLike = ( { dispatch }, { siteId, postId }, next, error ) => {
+	// skip data-layer middleware
+	next( {
+		type: UNLIKE_POST,
+		siteId,
+		postId,
+	} );
+}
+
+export default {
+	// watch for this action       -> initiate, onSuccess,  onFailure
+	[ LIKE_POST ]: [ dispatchRequest( likePost, verifyLike, undoLike ) ]
```

Comment:
My first reaction to this is: why do I need to call a function with success and failure arguments if that's already defined in my dispatched `http` action from `likePost` ? We'll see as I continue reviewing whether this starts to become more clear 😄 

_Edit:_ Ok, I see now how `dispatchRequest` handles this, though when would I ever set the value of an initiator's `onSuccess` to anything other than the original action? Would it be any clearer or better to define `verifyLike` and `undoLike` as the respective `onSuccess` and `onFailure` parameters of the `http` action?